### PR TITLE
Mikhails/homework template

### PR DESCRIPTION
**Pull request by Mikhail-SHA-256**
This is just template implementation part of homework

### CODE REVIEW

File: course6/toyapp_mikhail/displaylib/vector2d.h [hunk 1]
```diff
@@ -1,26 +1,9 @@
...
+namespace Display {
 
-		Vector2D operator + (const Vector2D& v)
-		{
-			auto r = Vector2D(*this);
-			r.x += v.x;
-			r.y += v.y;
-			return r;
-		}
+using Vector2D = VectorCommon<double, 2>;
```
Comment (maciekgajewski) [inline course6/toyapp_mikhail/displaylib/vector2d.h]: I would say it's statistical overkill to have a separate source file for one line

File: course6/toyapp_mikhail/displaylib/vector_common.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
+#pragma once
+
+namespace Display{
+
+template <class TValue, int TDimentions>
+class VectorCommon
+{
+	// No constructor for vector without specification
```
Comment (maciekgajewski) [inline course6/toyapp_mikhail/displaylib/vector_common.h]: Simply provide no definition for primary template at all, then only specializations will be allowed. 
But in this case I was expecting an actual primary template that works with any size

File: course6/toyapp_mikhail/displaylib/vector_common.h [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	VectorCommon() = delete;
+};
+
+// Specification for 2 dimentional vector
+template <class TValue>
```
Comment (maciekgajewski) [inline course6/toyapp_mikhail/displaylib/vector_common.h]: no no no, this is not what I had in mind. The goal of this template should be to use vector of any size/type

**Review by maciekgajewski — CHANGES_REQUESTED** [review]
This is not what I expected. See other's solutions

File: course6/toyapp_mikhail/displaylib/vector2d.h (hunk 1 above)
Comment (Mikhail-SHA-256) [reply course6/toyapp_mikhail/displaylib/vector2d.h]: Fixed. Left just one file with template definitions

File: course6/toyapp_mikhail/displaylib/vector_common.h (hunk 3 above)
Comment (Mikhail-SHA-256) [reply course6/toyapp_mikhail/displaylib/vector_common.h]: Changed.
But I still want to use nice decorations with x,y and x,y,z to not change client code.
Using v.x, v.y in client code is cleaner than v[0], v[1].
For decorations I had to use inheritance. is it ok?

Comment (maciekgajewski) [reply course6/toyapp_mikhail/displaylib/vector_common.h]: Forget about ```v.x```, ```v.y```. C++ doesn't have the support for properties.
The canonical way is to use accessor methods: getters and setters.

You may implement ```v.x()```, ```v.y()``` in an inherited class, and they can be an alternative api to ```v[0]```.

You will have to change your client code one way or the other. Consider this a punishment for using public fields :)

Comment (Mikhail-SHA-256) [reply course6/toyapp_mikhail/displaylib/vector_common.h]: Ok

**Comment by Mikhail-SHA-256** [comment]
Add more fixes based on feedback

Separate commit for reformatting

File: course6/toyapp_mikhail/displaylib/vector_common.h [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+#include <array>
+
+namespace Display {
+
+template <class TValue, size_t TDimentions, class TDecoratedType = TValue>
+class VectorCommon
+{
+private:
+	using TSelf = VectorCommon<TValue, TDimentions>;
```
Comment (maciekgajewski) [inline course6/toyapp_mikhail/displaylib/vector_common.h]: You can just use ```VectorCommon``` inside of the class' definition.

File: course6/toyapp_mikhail/displaylib/vector_common.h [hunk 5]
```diff
@@ -0,0 +1,100 @@
+#pragma once
+
+#include <array>
+
+namespace Display {
+
+template <class TValue, size_t TDimentions, class TDecoratedType = TValue>
```
Comment (maciekgajewski) [inline course6/toyapp_mikhail/displaylib/vector_common.h]: the "T" in "TDimensions" doesn't styand for "type"? right?
The popular convention is to use ALL_CAPS for value template arguments. Just like for constants; they are compile-time constants after all.

Comment (maciekgajewski) [reply course6/toyapp_mikhail/displaylib/vector_common.h]: Why is "TDecoratedType"? It's not imemdiately obvious for mew why would you need that in a vector, which means it needs eithjer better name, or at least explanation in a comment.

Comment (maciekgajewski) [reply course6/toyapp_mikhail/displaylib/vector_common.h]: Ah! Decorated type is the implementation! You are doing CRTP! Then using TValue as default makes no sense, and TSelf is useless.

File: course6/toyapp_mikhail/displaylib/vector_common.h [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+		(*this)[1] = yy;
+		(*this)[2] = zz;
+	}
+
+	Vector3DimDecorator(const TBase& original) :
+		Vector3DimDecorator(original[0], original[1], original[2])
+	{}
+};
+
+using Vector2D = Vector2DimDecorator<double>;
```
Comment (maciekgajewski) [inline course6/toyapp_mikhail/displaylib/vector_common.h]: Decorator/Decorated is not a canonical naming convention.
```Vector2Dim``` would be a better name, and ```VectorCommon``` should rather be ```VectorBase```

**Review by maciekgajewski — CHANGES_REQUESTED** [review]
Good use of CRTP!
But the naming and some details make it confusing.

File: course6/toyapp_mikhail/displaylib/vector_common.h (hunk 5 above)
Comment (Mikhail-SHA-256) [reply course6/toyapp_mikhail/displaylib/vector_common.h]: namespace Display::Detail

**Review by maciekgajewski — APPROVED** [review]
Good!

**Pull request merged**